File: web/src/engine/analysis/postSessionAnalyzer.ts
// Portage de Analysis/PostSessionAnalyzer.swift.
import type { AthleteProfile } from '../domain/profile';
import type { CompletedActivity } from '../domain/activity';
import type { Sport } from '../domain/types';
import { trainingPaces } from '../predictions/vdot';
import { LoadCalculator } from '../load/loadModel';

export interface SessionAnalysis {
  headline: string;
  insights: string[];
}

function sportLabel(s: Sport): string {
  switch (s) {
    case 'swim': return 'Natation';
    case 'bike': return 'Sortie vélo';
    case 'run': return 'Course';
    case 'strength': return 'Renforcement';
    case 'brick': return 'Brick';
  }
}

function durationText(s: number): string {
  const t = Math.round(s);
  const h = Math.trunc(t / 3600);
  const m = Math.trunc((t % 3600) / 60);
  return h > 0 ? `${h} h ${String(m).padStart(2, '0')}` : `${m} min`;
}
function kmText(m: number): string {
  return `${(m / 1000).toFixed(1)} km`;
}

export class PostSessionAnalyzer {
  analyze(a: CompletedActivity, profile: AthleteProfile, history: CompletedActivity[]): SessionAnalysis {
    const insights: string[] = [];
    const intensity = this.relativeIntensity(a, profile);

    let zoneWord: string;
    if (intensity == null) zoneWord = 'réalisée';
    else if (intensity >= 0.98) zoneWord = 'intense, proche du seuil';
    else if (intensity >= 0.88) zoneWord = 'au tempo';
    else if (intensity >= 0.75) zoneWord = 'en endurance active';
    else zoneWord = 'en endurance fondamentale';

    const headline = `${sportLabel(a.sport)} ${zoneWord} — ${durationText(a.duration)}${a.distanceM != null ? `, ${kmText(a.distanceM)}` : ''}.`;

    if (a.hrDriftPct != null) {
      const drift = a.hrDriftPct;
      if (drift < 5) insights.push(`Dérive cardiaque faible (${Math.trunc(drift)} %) : bonne endurance aérobie, effort bien tenu.`);
      else if (drift < 8) insights.push(`Dérive cardiaque modérée (${Math.trunc(drift)} %).`);
      else insights.push(`Dérive cardiaque élevée (${Math.trunc(drift)} %) : fatigue, chaleur ou allure trop rapide en début de séance.`);
    }

    const sameSport = history.filter((x) => x.sport === a.sport && x.id !== a.id);
    const trend = this.paceTrend(a, sameSport);
    if (trend) insights.push(trend);

    if (a.distanceM != null) {
      const prevDistances = sameSport.map((x) => x.distanceM).filter((x): x is number => x != null);
      if (prevDistances.length > 0 && a.distanceM > Math.max(...prevDistances)) {
        insights.push(`C'est ta plus longue ${sportLabel(a.sport).toLowerCase()} enregistrée 💪.`);
      }
    }

    const load = new LoadCalculator().load(a, profile);
    insights.push(`Charge estimée : ${Math.round(load)}.`);

    if (insights.length === 0) insights.push('Séance enregistrée. Continue comme ça !');
    return { headline, insights };
  }

  private relativeIntensity(a: CompletedActivity, profile: AthleteProfile): number | null {
    switch (a.sport) {
      case 'bike': {
        const p = a.normalizedPowerW ?? a.avgPowerW;
        if (p == null || profile.ftpWatts == null || profile.ftpWatts <= 0) return null;
        return p / profile.ftpWatts;
      }
      case 'run': {
        if (a.avgPaceSecPerKm == null || a.avgPaceSecPerKm <= 0 || profile.vdot == null) return null;
        const thr = trainingPaces(profile.vdot).thresholdSecPerKm;
        return 1000.0 / a.avgPaceSecPerKm / (1000.0 / thr);
      }
      case 'swim': {
        if (a.distanceM == null || a.distanceM <= 0 || profile.cssSecPer100m == null) return null;
        const speed = a.distanceM / a.duration;
        return speed / (100.0 / profile.cssSecPer100m);
      }
      default:
        return null;
    }
  }

  private paceTrend(a: CompletedActivity, sameSport: CompletedActivity[]): string | null {
    if (sameSport.length < 2) return null;
    if (a.sport === 'run') {
      const paces = sameSport.map((x) => x.avgPaceSecPerKm).filter((x): x is number => x != null);
      if (a.avgPaceSecPerKm == null || paces.length === 0) return null;
      const mean = paces.reduce((s, v) => s + v, 0) / paces.length;
      if (a.avgPaceSecPerKm < mean - 3) return `Allure plus rapide que ta moyenne récente : ta forme progresse (−${Math.trunc(mean - a.avgPaceSecPerKm)} s/km).`;
      if (a.avgPaceSecPerKm > mean + 6) return "Allure plus lente que d'habitude : séance facile ou fatigue.";
      return null;
    }
    if (a.sport === 'bike') {
      const powers = sameSport.map((x) => x.normalizedPowerW ?? x.avgPowerW).filter((x): x is number => x != null);
      const p = a.normalizedPowerW ?? a.avgPowerW;
      if (p == null || powers.length === 0) return null;
      const mean = powers.reduce((s, v) => s + v, 0) / powers.length;
      if (p > mean + 5) return 'Puissance au-dessus de ta moyenne récente : belle montée en forme.';
      return null;
    }
    return null;
  }
}

File: Packages/TriathlonEngine/Sources/TriathlonEngine/Analysis/PostSessionAnalyzer.swift
import Foundation

/// Résumé intelligent d'une séance réalisée, en langage naturel (français).
public struct SessionAnalysis: Sendable, Equatable {
    public var headline: String
    public var insights: [String]
    public init(headline: String, insights: [String]) {
        self.headline = headline
        self.insights = insights
    }
}

/// Analyse une activité réalisée : intensité, dérive cardiaque, tendance vs
/// l'historique, charge. Fonction pure, indépendante de l'UI.
public struct PostSessionAnalyzer: Sendable {
    public init() {}

    public func analyze(activity a: CompletedActivity, profile: AthleteProfile,
                        history: [CompletedActivity]) -> SessionAnalysis {
        var insights: [String] = []
        let intensity = relativeIntensity(a, profile: profile)

        // Intensité relative.
        let zoneWord: String
        switch intensity {
        case let x? where x >= 0.98: zoneWord = "intense, proche du seuil"
        case let x? where x >= 0.88: zoneWord = "au tempo"
        case let x? where x >= 0.75: zoneWord = "en endurance active"
        case .some: zoneWord = "en endurance fondamentale"
        case nil: zoneWord = "réalisée"
        }
        let headline = "\(sportLabel(a.sport)) \(zoneWord) — \(Self.duration(a.duration))\(a.distanceM.map { ", \(Self.km($0))" } ?? "")."

        // Dérive cardiaque.
        if let drift = a.hrDriftPct {
            if drift < 5 { insights.append("Dérive cardiaque faible (\(Int(drift)) %) : bonne endurance aérobie, effort bien tenu.") }
            else if drift < 8 { insights.append("Dérive cardiaque modérée (\(Int(drift)) %).") }
            else { insights.append("Dérive cardiaque élevée (\(Int(drift)) %) : fatigue, chaleur ou allure trop rapide en début de séance.") }
        }

        // Tendance vs historique (même sport).
        let sameSport = history.filter { $0.sport == a.sport && $0.id != a.id }
        if let trend = paceTrend(a, sameSport: sameSport) {
            insights.append(trend)
        }

        // Record de distance.
        if let dist = a.distanceM, let maxPrev = sameSport.compactMap(\.distanceM).max(), dist > maxPrev {
            insights.append("C'est ta plus longue \(sportLabel(a.sport).lowercased()) enregistrée 💪.")
        }

        // Charge.
        let load = LoadCalculator().load(for: a, profile: profile)
        insights.append("Charge estimée : \(Int(load.rounded())).")

        if insights.isEmpty { insights.append("Séance enregistrée. Continue comme ça !") }
        return SessionAnalysis(headline: headline, insights: insights)
    }

    // MARK: Intensité relative au seuil (0…~1.1)

    private func relativeIntensity(_ a: CompletedActivity, profile: AthleteProfile) -> Double? {
        switch a.sport {
        case .bike:
            guard let p = a.normalizedPowerW ?? a.avgPowerW, let ftp = profile.ftpWatts, ftp > 0 else { return nil }
            return Double(p) / Double(ftp)
        case .run:
            guard let pace = a.avgPaceSecPerKm, pace > 0, let vdot = profile.vdot else { return nil }
            let thr = VDOT.trainingPaces(vdot: vdot).thresholdSecPerKm
            return (1000.0 / pace) / (1000.0 / thr)   // rapport de vitesses
        case .swim:
            guard let dist = a.distanceM, dist > 0, let css = profile.cssSecPer100m else { return nil }
            let speed = dist / a.duration
            return speed / (100.0 / css)
        default:
            return nil
        }
    }

    private func paceTrend(_ a: CompletedActivity, sameSport: [CompletedActivity]) -> String? {
        guard sameSport.count >= 2 else { return nil }
        switch a.sport {
        case .run:
            let paces = sameSport.compactMap(\.avgPaceSecPerKm)
            guard let pace = a.avgPaceSecPerKm, !paces.isEmpty else { return nil }
            let mean = paces.reduce(0, +) / Double(paces.count)
            if pace < mean - 3 { return "Allure plus rapide que ta moyenne récente : ta forme progresse (−\(Int(mean - pace)) s/km)." }
            if pace > mean + 6 { return "Allure plus lente que d'habitude : séance facile ou fatigue." }
            return nil
        case .bike:
            let powers = sameSport.compactMap { $0.normalizedPowerW ?? $0.avgPowerW }
            guard let p = a.normalizedPowerW ?? a.avgPowerW, !powers.isEmpty else { return nil }
            let mean = Double(powers.reduce(0, +)) / Double(powers.count)
            if Double(p) > mean + 5 { return "Puissance au-dessus de ta moyenne récente : belle montée en forme." }
            return nil
        default:
            return nil
        }
    }

    // MARK: Formatage minimal (indépendant de l'app)

    static func duration(_ s: TimeInterval) -> String {
        let t = Int(s.rounded()); let h = t / 3600, m = (t % 3600) / 60
        return h > 0 ? "\(h) h \(String(format: "%02d", m))" : "\(m) min"
    }
    static func km(_ m: Double) -> String { String(format: "%.1f km", m / 1000) }

    private func sportLabel(_ s: Sport) -> String {
        switch s {
        case .swim: return "Natation"; case .bike: return "Sortie vélo"; case .run: return "Course"
        case .strength: return "Renforcement"; case .brick: return "Brick"
        }
    }
}
